File: javaprojects/client/src/main/java/minigames/client/MinigameNetworkClientWindow.java
package minigames.client;

import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JLayeredPane;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTextField;
import javax.swing.BoxLayout;
import javax.swing.JButton;

import minigames.client.backgrounds.Starfield;
import minigames.rendering.GameMetadata;
import minigames.rendering.GameServerDetails;

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Font;

import java.util.List;
import java.util.Optional;

/**
 * The main window that appears.
 * 
 * For simplicity, we give it a BorderLayout with panels for north, south, east, west, and center.
 * 
 * This makes it simpler for games to load up the UI however they wish, though the default expectation
 * is that the centre just has an 800x600 canvas.
 */
public class MinigameNetworkClientWindow {

    MinigameNetworkClient networkClient;

    JFrame frame;

    JPanel parent;
    JPanel north;
    JPanel center;
    JPanel south;
    JPanel west;
    JPanel east;    

    JLabel messageLabel;

    // We hang on to this one for registering in servers
    JTextField nameField;

    public MinigameNetworkClientWindow(MinigameNetworkClient networkClient) {
        this.networkClient = networkClient;

        frame = new JFrame();
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        parent = new JPanel(new BorderLayout());        

        north = new JPanel();
        parent.add(north, BorderLayout.NORTH);
        center = new JPanel();
        center.setPreferredSize(new Dimension(800, 600));
        parent.add(center, BorderLayout.CENTER);
        south = new JPanel();
        parent.add(south, BorderLayout.SOUTH);
        east = new JPanel();
        parent.add(east, BorderLayout.EAST);
        west = new JPanel();
        parent.add(west, BorderLayout.WEST);

        frame.add(parent);

        nameField = new JTextField(20);
        nameField.setText("Algernon");
    }

    /** Removes all components from the south panel */
    public void clearSouth() {
        south.removeAll();
    }

    /** Clears all sections of the UI  */
    public void clearAll() {
        for (JPanel p : new JPanel[] { north, south, east, west, center }) {
            p.removeAll();
        }
    }

    /** Adds a component to the north part of the main window */
    public void addNorth(java.awt.Component c) {
        north.add(c);
    }

    /** Adds a component to the south part of the main window */
    public void addSouth(java.awt.Component c) {
        south.add(c);
    }

    /** Adds a component to the east part of the main window */
    public void addEast(java.awt.Component c) {
        east.add(c);
    }

    /** Adds a component to the west part of the main window */
    public void addWest(java.awt.Component c) {
        west.add(c);
    }

    /** Adds a component to the center of the main window */
    public void addCenter(java.awt.Component c) {
        center.add(c);
    }

    /** "Packs" the frame, setting its size to match the preferred layout sizes of its component */
    public void pack() {
        frame.pack();
        parent.repaint();
    }

    /** Makes the main window visible */
    public void show() {
        pack();
        frame.setVisible(true);
    }

    /**
     * Shows a simple message layered over a retro-looking starfield.
     * Terrible placeholder art.
     */
    public void showStarfieldMessage(String s) {
        clearAll();

        JLayeredPane layeredPane = new JLayeredPane();
        layeredPane.add(new Starfield(networkClient.animator), JLayeredPane.DEFAULT_LAYER);
        layeredPane.setBackground(new Color(0,0,0,0));
        layeredPane.setPreferredSize(new Dimension(800, 600));

        JLabel label = new JLabel(s);
        label.setOpaque(true);
        label.setForeground(Color.CYAN);
        label.setBackground(Color.BLACK);
        label.setFont(new Font("Monospaced", Font.PLAIN, 36));
        Dimension labelSize = label.getPreferredSize();
        label.setSize(labelSize);
        label.setLocation((int)(400 - labelSize.getWidth() / 2), (int)(300 - labelSize.getHeight() / 2));
        layeredPane.add(label, JLayeredPane.MODAL_LAYER);

        center.add(layeredPane);
        pack();
    }

    /**
     * Shows a list of GameServers to pick from
     * 
     * TODO: Prettify!
     * @param servers
     */
    public void showGameServers(List<GameServerDetails> servers) {
        clearAll();

        JPanel panel = new JPanel();
        List<JPanel> serverPanels = servers.stream().map((gsd) -> {
            JPanel p = new JPanel();
            JLabel l = new JLabel(String.format("<html><h1>%s</h1><p>%s</p></html>", gsd.name(), gsd.description()));
            JButton newG = new JButton("Open games");

            newG.addActionListener((evt) -> {
                networkClient.getGameMetadata(gsd.name())
                  .onSuccess((list) -> showGames(gsd.name(), list));
            });

            p.add(l);
            p.add(newG);
            return p;
        }).toList();

        for (JPanel serverPanel : serverPanels) {
            panel.add(serverPanel);
        }

        center.add(panel);
        pack();
        parent.repaint();
    }

    /**
     * Shows a list of games to pick from
     * 
     * TODO: Prettify!
     * @param servers
     */
    public void showGames(String gameServer, List<GameMetadata> inProgress) {
        clearAll();

        JPanel namePanel = new JPanel();
        JLabel nameLabel = new JLabel("Your name");
        namePanel.add(nameLabel);
        namePanel.add(nameField);
        north.add(namePanel);


        JPanel panel = new JPanel();
        panel.setLayout(new BoxLayout(panel, BoxLayout.Y_AXIS));
        List<JPanel> gamePanels = inProgress.stream().map((g) -> {
            JPanel p = new JPanel();
            JLabel l = new JLabel(String.format("<html><h1>%s</h1><p>%s</p></html>", g.name(), String.join(",", g.players())));
            JButton join = new JButton("Join game");
            join.addActionListener((evt) -> {
                networkClient.joinGame(gameServer, g.name(), nameField.getText());
            });
            join.setEnabled(g.joinable());
            p.add(l);
            p.add(join);
            return p;
        }).toList();

        for (JPanel gamePanel : gamePanels) {
            panel.add(gamePanel);
        }

        JButton newG = new JButton("New game");
        newG.addActionListener((evt) -> {
            // FIXME: We've got a hardcoded player name here
            networkClient.newGame(gameServer, nameField.getText());
        });
        panel.add(newG);

        JScrollPane scrollPane = new JScrollPane(panel);
        scrollPane.setPreferredSize(new Dimension(800, 600));

        center.add(scrollPane);
        pack();
        parent.repaint();
    }


}

File: scalaprojects/scalafx-client/src/main/scala/minigames/scalafxclient/MinigameNetworkClientWindow.scala
package minigames.scalafxclient

import scalafx.scene.layout.*
import scalafx.scene.control.TextField
import scalafx.scene.Node
import scalafx.scene.Scene
import minigames.backgrounds.Starfield
import scalafx.scene.text.Text
import scalafx.geometry.VPos
import scalafx.scene.text.TextAlignment
import scalafx.scene.text.Font
import scalafx.scene.paint.Color

import minigames.rendering.*
import minigames.commands.*
import scalafx.scene.control.Label
import scalafx.scene.control.Button
import scalafx.application.Platform
import scalafx.geometry.Pos

/**
 * The main window that appears.
 * 
 * For simplicity, we give it a BorderLayout with panels for north, south, east, west, and center.
 * 
 * This makes it simpler for games to load up the UI however they wish, though the default expectation
 * is that the centre just has an 800x600 canvas.
 */
class MinigameNetworkClientWindow(val mgnClient: MinigameNetworkClient) {

    val north = HBox()
    val south = HBox()
    val east = HBox()
    val west = HBox()
    val center = HBox()

    def all = Seq(center, north, east, south, west)

    val borderPane = BorderPane(center, north, east, south, west)

    // We hang on to this one for registering in servers
    val nameField = new TextField { text = "Algernon" }

    def clearNorth():Unit = north.children.clear()
    def clearSouth():Unit = south.children.clear()
    def clearEast():Unit = east.children.clear()
    def clearWest():Unit = west.children.clear()
    def clearCenter():Unit = center.children.clear()

    def addNorth(node:Node) = north.children.append(node)
    def addSouth(node:Node) = south.children.append(node)
    def addEast(node:Node) = east.children.append(node)
    def addWest(node:Node) = west.children.append(node)
    def addCenter(node:Node) = center.children.append(node)

    def clearAll():Unit = for p <- all do p.children.clear()

    /** Makes the main window visible */
    def show():Unit = 
        App.stage.scene = new Scene {
            content = borderPane
        }
        App.stage.show()

    /** Shows a twinkling background with a label over it */
    def showStarfieldMessage(message:String):Unit = Platform.runLater {
        clearAll()
        addCenter(new StackPane {
            children = Seq(
                Starfield().canvas,
                new Text {
                    x = 400
                    y = 300
                    textOrigin = VPos.CENTER
                    textAlignment = TextAlignment.CENTER
                    text = message
                    font = Font.apply("Monospaced", 36)
                    fill = Color.Cyan
                }
            )
        })
        App.stage.sizeToScene()
        
    }

    /** Shows the list of game servers available for this client */
    def showGameServers(servers:Seq[GameServerDetails]):Unit = Platform.runLater {
        clearAll()
        addCenter(new VBox {
            alignment = Pos.CENTER
            children = 
                for s <- servers yield
                    new VBox {
                        children = Seq(
                            new Label {
                                text = s.name
                            },
                            new Label {
                                text = s.description
                            },
                            new Button {
                                text = "Open games"
                                onAction = (_) => {
                                    mgnClient.getGameMetadata(s.name).onSuccess((l) => showGames(s.name, l))
                                }
                            }
                        )
                    }
        })

    }

    def showGames(gameServer:String, games:Seq[GameMetadata]) = Platform.runLater {
        clearAll()
        addNorth(new HBox {
            children = Seq(
                new Label { text = "Your name " },
                nameField                
            )
        })
        addCenter(new VBox {
            alignment = Pos.CENTER
            children = 
                (for g <- games if g.joinable yield
                    new VBox {
                        children = Seq(
                            new Label {
                                text = g.name
                            },
                            new Label {
                                text = g.players.mkString(", ")
                            },
                            new Button {
                                text = "Join game"
                                onAction = (_) => {
                                    mgnClient.joinGame(gameServer, g.name, nameField.getText())
                                }
                            }
                        )
                    }
                ) :+ new VBox {
                    children = Seq(
                        new Button {
                            text = "New game"
                            onAction = (_) => {
                                mgnClient.newGame(gameServer, nameField.getText())
                            }
                        }
                    )
                }
        })
    }


}